src/test/error.ts

```diff
@@ -0,0 +1,32 @@
+import { expect } from "chai";
+import { FirebaseError } from "../error";
+
+describe("error", () => {
+  describe("FirebaseError", () => {
+    it("should be an instance of Error", () => {
+      const error = new FirebaseError("test-message");
+
+      expect(error).to.be.instanceOf(Error);
+    });
+
+    it("should apply default options", () => {
+      const error = new FirebaseError("test-message");
+
+      expect(error).to.deep.include({ children: [], exit: 1, name: "FirebaseError", status: 500 });
+    });
+
+    it("should persist all options", () => {
+      const allOptions: Required<ConstructorParameters<typeof FirebaseError>[1]> = {
```

Comment:
this `Required<ConstructorParameters<typeof FirebaseError>[1]>` statement seems so complex I'm almost certain it can't be correct.

- (a) does this object need a type? (does TS not catch that fields would be missing or incorrect when it tries to compile/run this?)
- (b) what is the advantage of all these typings? Why not export the `FirebaseErrorOptions` interface and use that?

Reply:
a) This type definition ensures that all possible options are specified in the object. That’s the purpose of this test, to check if all properties are persisted. If not for this type, TypeScript would allow an object with just one property specified, which would make the test incomplete. What’s incorrect about it?

b) There’s no reason in the source code for this interface to be exported. Usually it’s a good idea to keep APIs as small as possible, in this case only the FirebaseError class needs to be exported. Source code shouldn’t be influenced by tests.

Reply:
Ah, okay. I'm following your intentions a bit better now, thanks. I've sat thinking about this for a while and I think what doesn't align in my mind is that the tests _shouldn't_ be doing things that aren't done (or are intended to be done) in code. This typing (`Required...`) is incredibly ugly to me, and I would loath it if it started coming up around the code base. However, I see that your intention here is to ensure that _all_ of the parameters are defined and _all_ of them are included on the FirebaseError object. (If we _were_ wanting to type these options objects all over the place, then I'd make a much stronger case for exporting `FirebaseErrorOptions` and using that here, rather than `ConstructorParameters<...`)

So! Let's definitely keep this around. Would you please add a comment above `allOptions` to the effect of `// Typing here in order to enforce all parameters are defined.` so that it prompts the reader to understand why this (relatively complex) typing is here? That should help solidify the understanding of the test for the reader.

Reply:
Clear, I spit the definition in two so it's easier to understand each part and added appropriate comments, should be clear enough: b1bbe9b95279a7fa89dccb5fb3e4a87faf22cc2c